### PR TITLE
Add Call Stack graph

### PR DESCRIPTION
**Pull request by weilonge**
Add Call Stack graph and function-end log.

### CODE REVIEW

**Comment by dwi2** [comment]
@begeeben Please help to review this

File: lib/parser.js [hunk 1]
```diff
@@ -225,15 +237,19 @@ var parser = {
                 } else if (node.parent && node.parent.type === 'VariableDeclarator') {
                   functionName = node.parent.id.name;
                 }
-                var fnoutLine = '\nconsole.log("debuguy,"' +
+                var fnoutLine = 'var DEBOUT="debuguy,"' +
```
Comment (begeeben) [inline lib/parser.js]: Sorry that I couldn't find the definition of 'DEBOUT', what does it mean actually?

Comment (weilonge) [inline lib/parser.js]: Please see the line @244 & @245
node.loc.start.line + '";console.log(DEBOUT + ",STR");';
var fnEndLine = 'console.log(DEBOUT + ",END");';

**Comment by weilonge** [comment]
@begeeben I made rebase to my patch. Could you try to review it again? Thank you~

**Comment by weilonge** [comment]
@begeeben I made rebase to my patch. Could you try to review it again? Thank you~

File: cli.js [hunk 2]
```diff
@@ -3,10 +3,12 @@
 /**
  * Parsing:
  *
- * $ debuguy parse <source_dir>
- * $ debuguy parse <source_dir> <destination_dir>
- * $ debuguy parse -r <source_dir>
- * $ debuguy parse -r <source_dir> <destination_dir>
+ * $ debuguy parse <source_dir> [options]
+ * $ debuguy parse <source_dir> <destination_dir> [options]
```
Comment (dwi2) [inline cli.js]: You change the order of `options` here, but the option parsing logic is still the same.

Comment (dwi2) [inline cli.js]: I try the following command with your patch, we'll get `TypeError`
```sh
$> debuguy parse ./ -r
fs.js:661
  binding.readdir(pathModule._makeLong(path), callback);
          ^
TypeError: path must be a string
    at Object.fs.readdir (fs.js:661:11)
    at Object.pr_traverse [as traverse] (/Users/dwi2/Mozilla/debuguy/lib/parser.js:122:8)
    at Object.pr_parse [as parse] (/Users/dwi2/Mozilla/debuguy/lib/parser.js:170:10)
    at Object.module.exports.run (/Users/dwi2/Mozilla/debuguy/lib/app.js:79:18)
    at Object.<anonymous> (/Users/dwi2/Mozilla/debuguy/cli.js:25:5)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
```

Comment (dwi2) [inline cli.js]: But it works fine with the following command
```sh
$> debuguy parse -r ./
```

**Comment by dwi2** [comment]
I think the issue above should be fixed before landing. The description in `cli.js` is misleading.

File: cli.js (hunk 2 above)
Comment (weilonge) [inline cli.js]: I want to discuss [options] sequence actually. :)
Do you agree we move [options] to the last arguments?
If so, I will change the logic of these two lines in app.js
      var sourceDir = args.r ? args.r : args._[1];
      var destDir = args.r ? args._[1] : args._[2];

-r won't affect sourceDir and destDir anymore.

Comment (begeeben) [inline cli.js]: I suggest we follow the general consensus in using command line options, such as git. Making the sequence of the command line options different from such popular tool will confuse users.

In git, 'git checkout -b test_branch' works while 'git checkout test_branch -b' doesn't.

Comment (begeeben) [inline cli.js]: Would be good if it works either way.

File: lib/app.js [hunk 3]
```diff
@@ -41,11 +45,17 @@ module.exports = {
...
       var command = args._[0];
-      // XXX: This is tricky. Because minimist treat first params after
-      // -r as value of r. But it is actually sourceDir to us.
-      var sourceDir = args.r ? args.r : args._[1];
-      var destDir = args.r ? args._[1] : args._[2];
-      var parseOption;
+      var sourceDir = null, destDir = null, parseOption;
+
+      console.log(args._.length);
+      console.log(args._);
```
Comment (dwi2) [inline lib/app.js]: I think these two lines are for debugging purpose, right? If so, they should be removed.

File: lib/app.js [hunk 4]
```diff
@@ -41,11 +45,17 @@ module.exports = {
 
     if (args._ && args._.length > 0) {
       var command = args._[0];
-      // XXX: This is tricky. Because minimist treat first params after
-      // -r as value of r. But it is actually sourceDir to us.
-      var sourceDir = args.r ? args.r : args._[1];
-      var destDir = args.r ? args._[1] : args._[2];
-      var parseOption;
+      var sourceDir = null, destDir = null, parseOption;
```
Comment (dwi2) [inline lib/app.js]: I think it is better to leave them unassigned like this:
```js
var sourceDir, destDir, parseOption;
```
Every unassigned variable equals `undefined`.

**Comment by dwi2** [comment]
It looks good to me with just few nits. Please fix them before merging. Thanks

**Pull request merged**